### PR TITLE
Update tensor.py -> fix severe memory leaking issue!

### PR DESCRIPTION
**Pull request by HectorHHZ**
The requires_grad_ function does not respect the input flag x. This means no matter I intend to require gradients or not, histories are preserved and gradients will be computed. This can lead to severe memory "leak" over the training process since the system never releases unnecessary tensors. bug fixed

### CODE REVIEW

**Comment by HectorHHZ** [comment]
the bug leads to the severe memory usage in assignment 3 and assignment 2, the RAM usage will easily get hundreds of gigabytes during training.
